### ISSUE TITLE
[Bug] Connected Peer Metric doesn't count correctly.

### ISSUE THREAD

**Issue by andrewbaine** [libra/libra#3128]
# 🐛 Bug

On the premainnet network we saw connected peers grow in sawtooth from 9 to 40 while I believe the actual connected peers was never more than 9.

## To reproduce

You might be able to trigger this by repeatedly restarting a validator. We (Bison Trails) are happy to coordinate and trigger repeated restart of our validator.

**Stack trace/error message**

![image](https://user-images.githubusercontent.com/101534/77577781-c57e2e80-6ead-11ea-8f13-d8e9b8ca4623.png)

## Expected Behavior

We think that there was a node periodically restating, so the above graph should have just oscillated between 8 and 9.

## System information

**Please complete the following information:**

The above graph is from Bison Trails' validator running Libra 0.7.

## Additional context

None

**Comment by mgorven** [comment]
@phlip9 @bothra90

**Comment by bothra90** [comment]
Although small, this bug is triggered due to the incredibly complex way we handle disconnect events and requests in our system. Here's detailed write-up of what the core problem is, the current solution and it's short-comings, and a couple of possible solutions.

**tl;dr**: I'm proposing that we introduce a notion of connection identifiers to uniquely identify each connection. Passing this identifier around would allow us to disambiguate between different events across the system.

---
# Managing disconnections in Libra

## Goal

Deal with concurrent disconnection requests and events in the system, while satisfying the requirements discussed below

### Definitions

- We define a connection as "active" if it can be used for sending and receiving network messages.
- A central actor is responsible for listening for inbound connections and making outbound connections. We call this actor  the "PeerManager"
- Once a connection is correctly established, the PeerManager hands it over to another actor for handling I/O over it. In addition to I/O, this process is also responsible for disconnecting (when requested or on I/O errors) and reporting this back to the PeerManager as a `Disconnected` event. We call this actor the "Peer".

### Requirements

- We want to have at most 1 active connection per remote peer - we deal with this by having a deterministic protocol for deciding which connection to drop when we have more than 1 connection with a peer. If both end-points follow the same protocol, they should (mostly) drop the same connection.
- `LIBRA_NETWORK_PEERS` counter: We want to maintain and report a counter of "connected peers" that reflects the number of nodes we are connected to.
- Sending ACK on disconnect requests: If the upstream clients requests that we disconnect from a peer, we want to terminate the active connection and send an ACK when that is done.
- Notifying upstream client about `NewPeer`/ `LostPeer` events: We want to notify upstream clients with a `NewPeer` event when we first establish connectivity with a node and with a `LostPeer` event when we no longer have an active connection to the node. It is important to note that `NewPeer` and `LostPeer` events are non-commutative, i.e. delivering `NewPeer` after `LostPeer` has a different meaning from delivering `LostPeer` after `NewPeer`.

### Problem statement

There are 3 primary causes for a disconnect:

1. Simultaneous dial has occurred and one of the connections must be terminated (PeerManager-triggered)
2. Application requested a disconnect (client-triggered)
3. I/O error when trying to send or recv messages (Peer-triggered)

We want to tie a `Disconnected` event received by the PeerManager to one of the 3 causes above. We also want to determine if the event is for the currently active connection. Doing so will enable us to satisfy the requirements described above.

## Solutions:


### Current solution (buggy)

For the counter:

- We go a long way to ensure that the `LIBRA_NETWORK_PEERS` counter reflects the number of peers we are connected to without any over-counting. If, for example, we temporarily have 2 connections with a peer, we want to only count it once. This introduces a lot of complexity, eventually leading to bugs such as the one reported in [https://github.com/libra/libra/issues/3128](https://github.com/libra/libra/issues/3128).

For sending `LostPeer` notifications and/or ACKs to clients requesting disconnect:

- We use connection origin (inbound vs outbound) and disconnect reason (requested vs I/O error) to determine whether to notify upstream clients. If a connection is terminated because it was requested, we don't send `LostPeer` notifications or decrement the `LIBRA_NETWORK_PEERS` counter (this is the bug!). This is generally hard to get right, since we don't really know which connection we wanted to close and which the `Disconnect` notification corresponds to.

### Hacky fix

 A simple and hacky solution that satisfies the above requirements, but in an "eventually consistent" fashion is possible.

For the counter:

- Increment `LIBRA_NETWORK_PEERS` counter on each new inbound/outbound connection and decrement it on closing a connection. This means that we might intermittently see values of this counter not matching the number of unique peers we are connected to, but it should always stabilize at the correct value.

For sending `LostPeer` notifications and/or ACKs to clients requesting disconnect:

- Keep a count of the number of connections we have with any peer.  Whenever the counter drops to 0, cleanup any active state for the peer and notify upstream clients about `LostPeer`. This might cause some delay in delivery of `LostPeer` notification if the active connection was not the one terminated last.
- If upstream client had requested disconnecting from a peer, notify it on the first `Disconnected` event. This might lead to an early delivery of the ACK since the first event may not be from the termination of the active connection.

### Principled approach

We introduce the concept of a connection identifier (`cnxn_id`) that is assigned to each inbound/outbound connection. This `cnxn_id` is passed to Peer along with the connection. When the connection terminates, Peer sends the `cnxn_id` back to PeerManager along with the `Disconnected` event.

- On receiving a client request to disconnect, we move the active `cnxn_id` to a map from `cnxn_id` to the channel over which it can deliver the ACK while cleaning up any other state for the active connection. In reality, this will be a `HashMap<ConnectionId, Option<oneshot::Sender<Result()>>>` and we set the value to `None` when the disconnect is PeerManager triggered (to mitigate simultaneous dials)
- On receiving a `Disconnected` event, the PeerManager can unambiguously determine whether the lost connection was active or was in a pending disconnect state.
- If the active connection is lost, it sends a `LostPeer` notification to upstream clients. It can also decrement the `LIBRA_NETWORK_PEERS` counter here. Instead, if the terminated connection was pending disconnect state, the PeerManager removes the corresponding entry from the list of pending disconnects and sends an ACK to the client.

**Comment by davidiw** [comment]
Could we potentially leverage the PeerManager::active_peers::len? Whenever we add or remove a peer from this, we could also set the counter?

**Comment by bothra90** [comment]
@davidiw: Yes, that is effectively what would happen in the proposed (3rd) approach.

**Comment by bothra90** [comment]
Closed by #3200 

@mgorven can confirm when we deploy next to pre-mainnet

**Comment by mgorven** [comment]
This is happening again on pre-mainnet running 718ace8. Logs from Calibra node: P129173589
![firefox-20200416-151850](https://user-images.githubusercontent.com/183923/79512194-c6f6ce80-7ff5-11ea-86ac-158cb5cdd241.png)

**Comment by phlip9** [comment]
Closed by #3683
